crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -838,16 +838,36 @@ def upgrade_controller_clusters
 
         return upgrade_controllers_disruptive if upgrade_mode == :normal
 
-        network_node = ::Node.find(
-          "pacemaker_founder:true AND " \
+        network_cluster_members = ::Node.find(
+          "run_list_map:pacemaker-cluster-member AND " \
           "run_list_map:neutron-network AND NOT " \
           "run_list_map:neutron-server"
-        ).first
+        )
+
+        network_node = if network_cluster_members.empty?
+          nil
+        else
+          ::Node.load(network_cluster_members.first[:pacemaker][:founder])
+        end
+
         prepare_network_node(network_node) unless network_node.nil?
 
         # Now we must upgrade the clusters in the correct order:
         # 1. data, 2. API, 3. network
-        cluster_founders = ::Node.find("pacemaker_founder:true")
+
+        # search in all the cluster members to get a unique list of founders for all clusters
+        cluster_founders_names = ::Node.find("run_list_map:pacemaker-cluster-member").map! do |node|
+          node[:pacemaker][:founder]
+        end.uniq
+
+        # prepare the search for the founders
+        if cluster_founders_names.empty?
+          cluster_founders = []
+        else
+          search = "fqdn:"
+          search << cluster_founders_names.join(" OR fqdn:")
+          cluster_founders = ::Node.find(search)
+        end
```

Comment:
That feels quite complicated, why not:

```ruby
cluster_founders = cluster_founders_names.map { |name| ::Node.load(name) }
```

Reply:
becuase Im dumb probably, just guessing here :P